drivers/SmartThings/zigbee-window-treatment/src/aqara/init.lua

```diff
@@ -0,0 +1,258 @@
+  device:emit_event(initializedstate.initializedState.initializing())
+
+  device.thread:call_with_delay(2, function(d)
+    local lastLevel = device:get_latest_state("main", capabilities.windowShadeLevel.ID, capabilities.windowShadeLevel.shadeLevel.NAME) or 0
+    if lastLevel > 0 then
+      device:set_field(INIT_STATE, INIT_STATE_CLOSE)
+      send_close_cmd(device, command.component)
+    else
+      device:set_field(INIT_STATE, INIT_STATE_OPEN)
+      send_open_cmd(device, command.component)
+    end
+  end)
+end
+
+local function emit_level_event(device, shadeLevel)
+  if shadeLevel == 100 then
+    device:emit_event(capabilities.windowShade.windowShade.open())
+  elseif shadeLevel == 0 then
+    device:emit_event(capabilities.windowShade.windowShade.closed())
+  else
+    device:emit_event(capabilities.windowShade.windowShade.partially_open())
+  end
+end
+
+local function window_shade_level_cmd(driver, device, command)
+  local level = command.args.shadeLevel
+  device:send_to_component(command.component, WindowCovering.server.commands.GoToLiftPercentage(device, level))
+  device:emit_event(capabilities.windowShadeLevel.shadeLevel(level))
+end
+
+local function window_shade_open_cmd(driver, device, command)
+  local shadeLevel = device:get_field(SHADE_LEVEL) or 0
+  send_open_cmd(device, command.component)
+  emit_level_event(device, shadeLevel)
+end
+
+local function window_shade_close_cmd(driver, device, command)
+  local shadeLevel = device:get_field(SHADE_LEVEL) or 0
+  send_close_cmd(device, command.component)
+  emit_level_event(device, shadeLevel)
+end
+
+local function window_shade_pause_cmd(driver, device, command)
+  local shadeLevel = device:get_field(SHADE_LEVEL) or 0
+  device:send_to_component(command.component, WindowCovering.server.commands.Stop(device))
+  emit_level_event(device, shadeLevel)
+end
+
+local function motion_state_attr_handler(driver, device, value, zb_rx)
+  local state = value.value
+  device:set_field(SHADE_STATE, state)
+
+  if state == 0 then
+    local shadeLevel = device:get_field(SHADE_LEVEL) or 0
+    emit_level_event(device, shadeLevel)
+
+    local flag = device:get_field(INIT_STATE) or INIT_STATE_DONE
+    if flag == INIT_STATE_CLOSE then
+      device:set_field(INIT_STATE, INIT_STATE_REVERSE)
+      send_open_cmd(device, "main")
+    elseif flag == INIT_STATE_OPEN then
+      device:set_field(INIT_STATE, INIT_STATE_REVERSE)
+      send_close_cmd(device, "main")
+    elseif flag == INIT_STATE_REVERSE then
+      device:set_field(INIT_STATE, INIT_STATE_DONE)
+      read_pref_attribute(device)
+    end
+  elseif state == 1 then
+    device:emit_event(capabilities.windowShade.windowShade.opening())
+  elseif state == 2 then
+    device:emit_event(capabilities.windowShade.windowShade.closing())
+  end
+end
+
+local function current_position_attr_handler(driver, device, value, zb_rx)
+  local level = value.value
+  if level == 5.8774717541114e-39 then

```

Comment:
Why is this magic number here? This seems completely arbitrary.

Reply:
5.8774717541114e-39 is equal to 0. This code is to handle the data type exception. https://www.google.com/search?q=5.8774717541114e-39

Reply:
But the data type is Uint8 for this attribute, so there shouldn't be any situation where it would end up with this number here.

Reply:
A DataType of PresentValue attribute of AnalogOutput cluster is the SinglePrecisionFloat. https://developer-preview.smartthings.com/docs/edge-device-drivers/zigbee/generated/clusters/AnalogOutput.html?highlight=analogoutput#st.zigbee.zcl.clusters.AnalogOutput.PresentValue 

I paste my output log below.
---
2022-07-28T00:43:00.896433581+00:00 INFO Zigbee Window Treatment  <ZigbeeDevice: 61da00c9-778f-4d56-91a8-3e90937bba2f [0xC1A6] (Aqara Curtain Controller)> received Zigbee message: < ZigbeeMessageRx || type: 0x00, < AddressHeader || src_addr: 0xC1A6, src_endpoint: 0x01, dest_addr: 0x0000, dest_endpoint: 0x01, profile: 0x0104, cluster: AnalogOutput >, lqi: 0xC8, rssi: -50, body_length: 0x0011, < ZCLMessageBody || < ZCLHeader || frame_ctrl: 0x18, seqno: 0x15, ZCLCommandId: 0x0A >, < ReportAttribute || < AttributeRecord || AttributeId: 0x0055, DataType: SinglePrecisionFloat, PresentValue: 0 >, < AttributeRecord || AttributeId: 0xF000, DataType: Uint32, Uint32: 0x00000000 > > > >
2022-07-28T00:43:00.924165581+00:00 TRACE Zigbee Window Treatment  Found ZigbeeMessageDispatcher handler in zigbee_window_treatment -> Aqara Window Treatment Handler -> Aqara Window Treatment Handler
2022-07-28T00:43:00.930224914+00:00 INFO Zigbee Window Treatment  Executing ZclClusterAttributeValueHandler: cluster: AnalogOutput, attribute: PresentValue
2022-07-28T00:43:00.937048581+00:00 INFO Zigbee Window Treatment  ---------> current_position_attr_handler value.value: 5.8774717541114e-39
2022-07-28T00:43:00.943240247+00:00 INFO Zigbee Window Treatment  <ZigbeeDevice: 61da00c9-778f-4d56-91a8-3e90937bba2f [0xC1A6] (Aqara Curtain Controller)> emitting event: {"attribute_id":"shadeLevel","capability_id":"windowShadeLevel","component_id":"main","state":{"value":0}}
2022-07-28T00:43:00.959368581+00:00 INFO Zigbee Window Treatment  <ZigbeeDevice: 61da00c9-778f-4d56-91a8-3e90937bba2f [0xC1A6] (Aqara Curtain Controller)> emitting event: {"attribute_id":"windowShade","capability_id":"windowShade","component_id":"main","state":{"value":"closed"}}

Reply:
In any case this is a bug, I will create a JIRA so this doesn't happen.  In addition, doesn't the `round` call below convert this value to 0 anyway?

Reply:
I checked and updated the code. Thanks for your comments :)